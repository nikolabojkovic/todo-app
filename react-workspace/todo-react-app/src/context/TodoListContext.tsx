import { createContext, useContext, useReducer } from "react";

import { IAction, TodoActions, IPaging, ITodo, StateFilter, ISettings } from "../models";
import { IState } from "./";
 
export const TodosContext = createContext({} as IState);
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const TodosDispatchContext = createContext(null as any);

type Props = {
  children: JSX.Element | JSX.Element[],
  initialState: IState
}

export function TodoStateProvider({ initialState, children }: Props) {
  const [state, dispatch] = useReducer(
    todoStateReducer,
    initialState
  );

  return (
    <TodosContext.Provider value={state}>
      <TodosDispatchContext.Provider value={dispatch}>
        {children}
      </TodosDispatchContext.Provider>
    </TodosContext.Provider>
  );
}

export function useTodoList() {
  return useContext(TodosContext);
}

export function useTodoListDispatch() {
  return useContext(TodosDispatchContext);
}

export default TodosDispatchContext.Provider;

export function todoStateReducer(state: IState, action: IAction) {
  switch (action.type) { 
    case TodoActions.loadingStarted: {
      return { 
        ...state,
        effectTrigger: null,
        isLoading: true
      } as IState;
    }
    case TodoActions.activeTabChanged: {
      return { 
        ...state,
        activeTab: action.payload.activeTab
      } as IState;
    }
    case TodoActions.fetch: {
      return { 
        ...state,
        effectTrigger: { type: TodoActions.fetch, payload: action.payload },
      } as IState;
    }
    case TodoActions.filter: {
      return { 
        ...state,
        effectTrigger: { type: TodoActions.filter, payload: action.payload },
      } as IState;
    }
    case TodoActions.search: {
      return { 
        ...state,
        effectTrigger: { type: TodoActions.search, payload: action.payload },
      } as IState;
    }
    case TodoActions.sort: {
      return { 
        ...state,
        effectTrigger: { type: TodoActions.sort, payload: action.payload },
      } as IState;
    }

    case TodoActions.fetched: {
      return {
        ...state,
        isLoading: false,
        originalList: action.payload.list,
        displayList: action.payload.list,
        effectTrigger: null,
        paging: {
          ...state.paging,
          totalCount: action.payload.list.length,
          activePage: state.paging.activePage === 0 ? 1 : state.paging.activePage,
          itemsPerPage: state.paging.itemsPerPage,
          startIndex: ((state.paging.activePage === 0 ? 1 : state.paging.activePage) - 1) * state.paging.itemsPerPage,
          endIndex: (state.paging.activePage === 0 ? 1 : state.paging.activePage) * state.paging.itemsPerPage
        } as IPaging,
      };
    }
    case TodoActions.searched: {
      return {
        ...state,
        isLoading: false,
        effectTrigger: null,
        displayList: [...action.payload.list],
        paging: {
          ...state.paging,
          activePage: action.payload.activePage,
          totalCount: action.payload.list.length,
          startIndex: (action.payload.activePage - 1) * state.paging.itemsPerPage,
          endIndex: action.payload.activePage * state.paging.itemsPerPage
        } as IPaging
      };
    }
    case TodoActions.filtered: {
      return {
        ...state,
        isLoading: false,
        effectTrigger: null,
        displayList: [...action.payload.list],
        filter: {...action.payload.filter},
        paging: {
          ...state.paging,
          activePage: action.payload.activePage,
          totalCount: action.payload.list.length,
          startIndex: (action.payload.activePage - 1) * state.paging.itemsPerPage,
          endIndex: action.payload.activePage * state.paging.itemsPerPage
        } as IPaging
      };
    }
    case TodoActions.sorted: {
      return {
        ...state,
        isLoading: false,
        effectTrigger: null,
        displayList: [...action.payload.list],
        sort: {...action.payload.sort},
        paging: {...state.paging} as IPaging
      };
    }
    case TodoActions.imported: {
      return {
        ...state,
        originalList: [...action.payload.list],
        displayList: [...action.payload.list],
        effectTrigger: { type: TodoActions.imported },
        search: { searchTerm: '' },
        filter: {
          state: StateFilter.all
        },
        paging: {
          ...state.paging,
          activePage: action.payload.activePage,
          totalCount: action.payload.list.length,
          startIndex: (action.payload.activePage - 1) * state.paging.itemsPerPage,
          endIndex: action.payload.activePage * state.paging.itemsPerPage
        } as IPaging
      };
    }

    case TodoActions.searchTermUpdated: {
      return {
        ...state,
        effectTrigger: null,
        search: { searchTerm: action.payload.searchTerm },
      };
    }    
    case TodoActions.pagingFatched: {
      return {
        ...state,
        effectTrigger: null,
        paging: { 
          ...state.paging,
          totalCount: action.payload.totalCount,
          activePage: action.payload.activePage,
          itemsPerPage: action.payload.itemsPerPage,
          startIndex: (action.payload.activePage - 1) * action.payload.itemsPerPage,
          endIndex: action.payload.activePage * action.payload.itemsPerPage
        } as IPaging
      } as IState;
    }
    case TodoActions.pagingUpdated: {
      return {
        ...state,
        effectTrigger: { type: TodoActions.pagingUpdated },
        paging: {
          ...state.paging,
          activePage: action.payload.activePage,
          itemsPerPage: action.payload.itemsPerPage,
          startIndex: (action.payload.activePage - 1) * action.payload.itemsPerPage,
          endIndex: action.payload.activePage * action.payload.itemsPerPage
        } as IPaging
      } as IState;
    }
    case TodoActions.settingsFetched: {
      return {
        ...state,
        effectTrigger: null,
        settings: {
          ...action.payload
        } as ISettings
      };
    }
    case TodoActions.settingsUpdated: {
      return {
        ...state,
        effectTrigger: { type: TodoActions.settingsUpdated },
        settings: {
          ...action.payload
        } as ISettings
      };
    }

    case TodoActions.added: {
      const id = state.originalList.length >= 1 
      ? state.originalList
          .sort((a: ITodo, b: ITodo) => a.id > b.id ? 1 : -1)[state.originalList.length - 1].id 
      : 0;
      const newTodo = {
        id: id + 1,
        title: action.payload.title,
        description: action.payload.description,
        completed: false, 
        createdAt: action.payload.createdAt
      } as ITodo;

      return {
        ...state,
        originalList: [...state.originalList, newTodo],
        displayList: [...state.displayList, newTodo],
        effectTrigger: { type: TodoActions.added },
        paging: {
          ...state.paging, 
          totalCount: state.paging.totalCount + 1,
          activePage: calculateActivePageOnAdd(state.paging),
          startIndex: (calculateActivePageOnAdd(state.paging) - 1) * state.paging.itemsPerPage,
          endIndex: calculateActivePageOnAdd(state.paging) * state.paging.itemsPerPage
        } as IPaging
      } as IState;
    }
    case TodoActions.changed: {
      return {
        ...state,
        originalList: state.originalList.map(t => {
          if (t.id === action.payload.todo.id) {
            return action.payload.todo;
          } else {
            return t;
          }
        }),
        displayList: state.displayList.map(t => {
          if (t.id === action.payload.todo.id) {
            return action.payload.todo;
          } else {
            return t;
          }
        }),
        effectTrigger: { type: TodoActions.changed },
        paging: {...state.paging}
      } as IState;
    }
    case TodoActions.deleted: {
      return {
        ...state,
        originalList: state.originalList.filter(t => t.id !== action.payload.id),
        displayList: state.displayList.filter(t => t.id !== action.payload.id),
        effectTrigger: { type: TodoActions.deleted },
        paging: {
          ...state.paging, 
          totalCount: state.paging.totalCount - 1,
          activePage: calculateActivePageOnDelete(state.paging),
          startIndex: (calculateActivePageOnDelete(state.paging) - 1) * state.paging.itemsPerPage,
          endIndex: calculateActivePageOnDelete(state.paging) * state.paging.itemsPerPage,  
        } as IPaging
      } as IState;
    }
    default: {
      throw Error('Unknown action: ' + action.type);
    }
  }
}

function calculateActivePageOnDelete(paging: IPaging) {
  return Math.ceil((paging.totalCount - 1) / paging.itemsPerPage) < paging.activePage
    ? paging.activePage - 1 : paging.activePage;
}

function calculateActivePageOnAdd(paging: IPaging) {
  return Math.ceil((paging.totalCount + 1) / paging.itemsPerPage) > paging.activePage
    ? Math.ceil((paging.totalCount + 1) / paging.itemsPerPage) : paging.activePage;
}