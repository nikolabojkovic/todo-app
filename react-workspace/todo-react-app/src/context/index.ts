export * from './IState';
export * from './TodoListContext';
export * from './testData';