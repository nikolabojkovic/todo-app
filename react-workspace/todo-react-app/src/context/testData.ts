import { ITodo } from "../models/Todo";
import { IPaging } from "../models/IPaging";
import { IState } from "./IState";
import { ISort, SortDirection } from "../models/ISort";
import { IFilter, StateFilter } from "../models/IFilter";
import { IGeneralSettings, IPaginationSettings, ISearchSettings, ISettings, ListContainerType, PaginationType } from "../models/ISettings";

export const todoListTestData: ITodo[] = [
  {
    id: 1,
    title: "Task 1",
    description: "Description 1",
    completed: false,
    createdAt: new Date(2022, 1, 4)
  } as ITodo,
  {
    id: 2,
    title: "Task 2",
    description: "Description 2",
    completed: true,
    createdAt: new Date(2022, 1, 5)
  } as ITodo,
  {
    id: 3,
    title: "Task 3",
    description: "Description 3",
    completed: false,
    createdAt: new Date(2022, 1, 6)
  } as ITodo,
  {
    id: 4,
    title: "Task 4",
    description: "Description 4",
    completed: false,
    createdAt: new Date(2022, 1, 7)
  } as ITodo,
  {
    id: 6,
    title: "Task 6",
    description: "Description 6",
    completed: false,
    createdAt: new Date(2022, 2, 4)
  } as ITodo,
  {
    id: 5,
    title: "Task 5",
    description: "Description 5",
    completed: false,
    createdAt: new Date(2022, 3, 8)
  } as ITodo
] as ITodo[];

export const stateTestData: IState = 
{
  originalList: todoListTestData,
  displayList: todoListTestData,
  search: { searchTerm: '' },
  filter: { state: StateFilter.all } as IFilter,
  sort: { column: 'createdAt', direction: SortDirection.Asc} as ISort,
  paging: {
    totalCount: 6,
    activePage: 1,
    itemsPerPage: 5,
    startIndex: 0,
    endIndex: 5,
  } as IPaging,
  settings: {
    general: {
      isConfirmEnabled: true,
      isPaginationEnabled: true,
      isInfiniteScrollEnabled: false,
      listSizeType: ListContainerType.Dynamic,
      fixedListSize: 200
    } as IGeneralSettings,
    search: {
      isSearchOnKeyPressEnabled: false,
      debounceTime: 500
    } as ISearchSettings,
    pagination: {
      paginationType: PaginationType.Classic,
      maxVisiblePages: 3
    } as IPaginationSettings
  } as ISettings
} as IState;