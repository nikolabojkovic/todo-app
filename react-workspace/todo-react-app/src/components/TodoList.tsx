import { useEffect } from 'react'
import { AddTodo } from './AddTodo';
import { TodoItem } from './TodoItem';
import { Paging } from './Paging';
import { Search } from './Search';
import { useTodoList } from '../context/TodosContext';
import { Tabs } from './Tabs';
import { FilterTodos } from './FilterTodos';
import { Sorting } from './Sorting';
import { ImportExport } from './ImportExport';
import { ITodo } from '../models/Todo';

export function TodoList() {
  const todoList = useTodoList();

  // TODO: move save to context
  useEffect(() => {
    localStorage.setItem('todo-list', JSON.stringify(todoList.originalList));
  }, [todoList.originalList]);

  return (
    <main className="App__todo-list">
      <Tabs>
        <AddTodo key="add-todo" />
        <Search 
          key="search-todos" 
          placeholder='Search by title or description' 
        />
        <FilterTodos key="filter-todos" />
        <ImportExport key="import-export" />
      </Tabs>
      <Sorting/>
      {todoList.displayList
        .slice(todoList.paging.startIndex, todoList.paging.endIndex)
        .map((todo: ITodo) => <TodoItem key={todo.id} todo={todo} />)}
      <Paging/>
    </main>
  );
}