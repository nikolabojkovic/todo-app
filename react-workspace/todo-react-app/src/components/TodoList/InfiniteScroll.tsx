import { ReactNode, useEffect, useState } from "react";
import { delay, Subject } from "rxjs";

import { useTodoList } from "../../context";
import { Loader, LoadMore } from "..";
import { ListContainerType } from "../../models";

type Props = {
  children: ReactNode,
  setEndIndex: (index: number) => void
}

export function InfiniteScroll({children, setEndIndex}: Props) {
  const todoList = useTodoList();

  const itemheight = 55;
  const [infiniteScroll, setInfiniteScroll] = useState({ 
    startIndex: 0, 
    endIndex: calculateinfiniteScrollEndIndex(),
    isLoading: false,
    fetch: new Subject<number>()
  });

  const hasMoreItemsToLoad = !infiniteScroll.isLoading && (infiniteScroll.endIndex < todoList.displayList.length);

  useEffect(() => {
    setEndIndex(calculateinfiniteScrollEndIndex());
    const subscription = infiniteScroll.fetch
      .pipe(
        delay(1000),
      )
      .subscribe((endIndex: number) => {
        const end = endIndex + 5;
        setInfiniteScroll({...infiniteScroll, endIndex: end, isLoading: false});
        setEndIndex(end);
      });

    return () => subscription.unsubscribe();
  }, []);

  useEffect(() => {
    window.addEventListener("scroll", handleScroll);
    return () => {
      window.removeEventListener("scroll", handleScroll);
    };
  });

  useEffect(() => {
    const endIndex = calculateinfiniteScrollEndIndex();
    setEndIndex(endIndex);
    setInfiniteScroll({
      startIndex: 0, 
      endIndex,
      isLoading: false,
      fetch: infiniteScroll.fetch
    });
  }, [todoList.settings.general.listSizeType]);

  const handleScroll = () => {
    if (infiniteScroll.isLoading) {
      return;
    }

    let { scrollTop, clientHeight, scrollHeight } = document.documentElement;

    if (todoList.settings.general.listSizeType === ListContainerType.Fixed) {
      scrollTop = document.getElementById('todo-list-container')!.scrollTop;
      clientHeight = document.getElementById('todo-list-container')!.clientHeight;      
      scrollHeight = document.getElementById('todo-list-container')!.scrollHeight;
    }

    if ((scrollTop + clientHeight >= scrollHeight - 20) 
     && !infiniteScroll.isLoading 
     && hasMoreItemsToLoad) {
      setInfiniteScroll({...infiniteScroll, isLoading: true });
      infiniteScroll.fetch.next(infiniteScroll.endIndex);
    }    
  };
  
  function calculateinfiniteScrollEndIndex(): number {
    const endIndex = (todoList.settings.general.listSizeType === ListContainerType.Fixed  
      ? ((todoList.settings.general.fixedListSize) / itemheight)
      : (document.documentElement.clientHeight - 200) / itemheight);

    return endIndex;
  }
  
  return (
    <div 
      onScroll={handleScroll}     
      style={ 
        todoList.settings.general.listSizeType === ListContainerType.Fixed 
        ? { 
            height: todoList.settings.general.fixedListSize,
            overflow: 'overlay'
          } : {}
        }
      
    >
      { children }
      { infiniteScroll.isLoading &&
        <Loader height={150} />
      }  
      { hasMoreItemsToLoad && 
        <LoadMore />
      } 
    </div>
  );
}