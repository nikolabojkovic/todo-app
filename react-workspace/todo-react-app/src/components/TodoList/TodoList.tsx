import { useEffect, useState } from 'react';
import { Subject, delay, first } from 'rxjs';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faAngleDown } from '@fortawesome/free-solid-svg-icons';

import { ISort, ITodo, IAction, TodoActions, ListContainerType } from '../../models';
import { useTodoList, useTodoListDispatch } from '../../context';
import { ITodoListProvider } from '../../providers';
import { Loader, TodoItem } from '../';
import { useTodoListEffect } from '../../hooks/UseTodoListEffects';

type Props = {
  todoListProvider: ITodoListProvider
};

export function TodoList({ todoListProvider }: Props) {
  const todoList = useTodoList();
  const dispatch = useTodoListDispatch();
  const todoItemheight = 55;

  const [infiniteScroll, setInfiniteScroll] = useState({ 
    startIndex: 0, 
    endIndex: calculateinfiniteScrollEndIndex(),
    isLoading: false,
    fetch: new Subject<number>()
  });

  useTodoListEffect(todoList, todoListProvider, dispatch);

  useEffect(() => { 
    dispatch({
      type: TodoActions.fetch,
      payload: {
        sort: {
          column: 'createdAt',
          direction: 'asc'
        } as ISort
      }
    } as IAction);
  }, [dispatch]);

  useEffect(() => {
    window.addEventListener("scroll", handleScroll);
    return () => {
      window.removeEventListener("scroll", handleScroll);
    };
  });

  useEffect(() => {
    const subscription = infiniteScroll.fetch
      .pipe(
        delay(1000), // TODO: switchMap to http request which whill load data from BE
      )
      .subscribe((endIndex: number) => {
        setInfiniteScroll({...infiniteScroll, endIndex: endIndex + 5, isLoading: false});
      });

    return () => subscription.unsubscribe();
  }, []);

  useEffect(() => {
    setInfiniteScroll({
      startIndex: 0, 
      endIndex: calculateinfiniteScrollEndIndex(),
      isLoading: false,
      fetch: infiniteScroll.fetch
    });
  }, [todoList.settings.general.listSizeType]);

  function calculateinfiniteScrollEndIndex(): number {
    return (todoList.settings.general.listSizeType === ListContainerType.Fixed  
      ? ((todoList.settings.general.fixedListSize) / todoItemheight)
      : (document.documentElement.clientHeight - 200) / todoItemheight);
  }

  function getDisplayList() {
    if (todoList.settings.general.isPaginationEnabled) {
      return todoList.displayList
        .slice(todoList.paging.startIndex, todoList.paging.endIndex)
        .map((todo: ITodo) => <TodoItem key={todo.id} todo={todo} />);
    }

    if (todoList.settings.general.isInfiniteScrollEnabled) {
      return todoList.displayList
        .slice(infiniteScroll.startIndex, infiniteScroll.endIndex)
        .map((todo: ITodo) => <TodoItem key={todo.id} todo={todo} />);
    }
      
    return todoList.displayList.map((todo: ITodo) => <TodoItem key={todo.id} todo={todo} />);
  }

  const handleScroll = () => {
    if (!todoList.settings.general.isInfiniteScrollEnabled || infiniteScroll.isLoading) {
      return;
    }

    let { scrollTop, clientHeight, scrollHeight } = document.documentElement;

    if (todoList.settings.general.listSizeType === ListContainerType.Fixed) {
      scrollTop = document.getElementById('todo-list-container')!.scrollTop;
      clientHeight = document.getElementById('todo-list-container')!.clientHeight;      
      scrollHeight = document.getElementById('todo-list-container')!.scrollHeight;
    }

    if ((scrollTop + clientHeight >= scrollHeight - 20) 
     && !infiniteScroll.isLoading 
     && infiniteScroll.endIndex < todoList.originalList.length) {
      setInfiniteScroll({...infiniteScroll, isLoading: true });
      infiniteScroll.fetch.next(infiniteScroll.endIndex);
    }    
  };

  return (
    <>
    { todoList.activeTab !== 'settings' && 
      <main 
        id="todo-list-container" 
        className="App__todo-list" 
        onScroll={handleScroll}     
        style={ 
          todoList.settings.general.listSizeType === ListContainerType.Fixed 
          ? { 
              height: todoList.settings.general.fixedListSize,
              overflow: 'overlay'
            } 
          : {}
          }
        >
        { 
          todoList.isLoading 
          ? <Loader height={280} />
          : <section 
              id="todo-list-section"
              className=''
            >
              {
                (todoList.paging.totalCount > 0 ?
                  getDisplayList() 
                : <div className='text-light mt-5 mb-5 fade-in'>No data</div>)
              }
              {
                todoList.settings.general.isInfiniteScrollEnabled &&
                infiniteScroll.isLoading &&
                <Loader height={150} />
              }              
            </section>
        }
        {
          todoList.settings.general.isInfiniteScrollEnabled &&
          !infiniteScroll.isLoading &&
          infiniteScroll.endIndex < todoList.originalList.length &&
          <div id="infinite-scroll-end" className="App__todo-list__infinite-scroll-end">
            <div className="App__todo-list__infinite-scroll-end--bouncing">
              <FontAwesomeIcon icon={faAngleDown} />
              <div>Scroll down to load more</div>
            </div>            
          </div>
        }      
      </main>
    }    
    </>
  );
}