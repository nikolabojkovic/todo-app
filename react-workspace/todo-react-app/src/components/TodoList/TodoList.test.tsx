import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import { TodoList } from './TodoList';
import { TodoStateProvider } from '../../context/TodoListContext';
import { stateTestData } from '../../context/testData';

describe('todo list rendered', () => {
  it('match snapshot', () => {
    const todoListJsxElements = (<TodoStateProvider todoList={stateTestData}>
                                   <TodoList/>
                                 </TodoStateProvider>);
    const tree = renderer.create(
      todoListJsxElements
    ).toJSON();
    render(todoListJsxElements);
    expect(tree).toMatchSnapshot();
  });
});
