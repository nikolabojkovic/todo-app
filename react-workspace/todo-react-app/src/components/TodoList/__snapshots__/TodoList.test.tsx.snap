// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo list rendered match snapshot list 1`] = `
<main
  className="App__todo-list"
  onScroll={[Function]}
  style={Object {}}
>
  <section
    className=""
    onScroll={[Function]}
  >
    <div
      className="App__todo-list__item"
    >
      <div
        className="hstack gap-3"
      >
        <div>
          <div
            className="App__todo-list__item-title"
          >
            Task 1
          </div>
          <div
            className="App__todo-list__item-description"
          >
            <span>
              Description 1
            </span>
            <span>
               (Fri Feb 04 2022)
            </span>
          </div>
        </div>
        <div
          className="ms-auto"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-check-double action-icon"
            data-icon="check-double"
            data-prefix="fas"
            data-testid="complete-button"
            focusable="false"
            onClick={[Function]}
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M374.6 86.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 178.7l-57.4-57.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l80 80c12.5 12.5 32.8 12.5 45.3 0l160-160zm96 128c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 402.7 86.6 297.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l128 128c12.5 12.5 32.8 12.5 45.3 0l256-256z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
        <div>
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-trash action-icon"
            data-icon="trash"
            data-prefix="fas"
            data-testid="delete-button"
            focusable="false"
            onClick={[Function]}
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="App__todo-list__item"
    >
      <div
        className="hstack gap-3"
      >
        <div>
          <div
            className="App__todo-list__item-title--completed"
          >
            Task 2
          </div>
          <div
            className="App__todo-list__item-description--compleated"
          >
            <span>
              Description 2
            </span>
            <span>
               (Sat Feb 05 2022)
            </span>
          </div>
        </div>
        <div
          className="ms-auto"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-check-double action-icon--disabled"
            data-icon="check-double"
            data-prefix="fas"
            data-testid="complete-button"
            focusable="false"
            onClick={[Function]}
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M374.6 86.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 178.7l-57.4-57.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l80 80c12.5 12.5 32.8 12.5 45.3 0l160-160zm96 128c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 402.7 86.6 297.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l128 128c12.5 12.5 32.8 12.5 45.3 0l256-256z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
        <div>
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-trash action-icon"
            data-icon="trash"
            data-prefix="fas"
            data-testid="delete-button"
            focusable="false"
            onClick={[Function]}
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="App__todo-list__item"
    >
      <div
        className="hstack gap-3"
      >
        <div>
          <div
            className="App__todo-list__item-title"
          >
            Task 3
          </div>
          <div
            className="App__todo-list__item-description"
          >
            <span>
              Description 3
            </span>
            <span>
               (Sun Feb 06 2022)
            </span>
          </div>
        </div>
        <div
          className="ms-auto"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-check-double action-icon"
            data-icon="check-double"
            data-prefix="fas"
            data-testid="complete-button"
            focusable="false"
            onClick={[Function]}
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M374.6 86.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 178.7l-57.4-57.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l80 80c12.5 12.5 32.8 12.5 45.3 0l160-160zm96 128c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 402.7 86.6 297.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l128 128c12.5 12.5 32.8 12.5 45.3 0l256-256z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
        <div>
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-trash action-icon"
            data-icon="trash"
            data-prefix="fas"
            data-testid="delete-button"
            focusable="false"
            onClick={[Function]}
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="App__todo-list__item"
    >
      <div
        className="hstack gap-3"
      >
        <div>
          <div
            className="App__todo-list__item-title"
          >
            Task 4
          </div>
          <div
            className="App__todo-list__item-description"
          >
            <span>
              Description 4
            </span>
            <span>
               (Mon Feb 07 2022)
            </span>
          </div>
        </div>
        <div
          className="ms-auto"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-check-double action-icon"
            data-icon="check-double"
            data-prefix="fas"
            data-testid="complete-button"
            focusable="false"
            onClick={[Function]}
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M374.6 86.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 178.7l-57.4-57.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l80 80c12.5 12.5 32.8 12.5 45.3 0l160-160zm96 128c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 402.7 86.6 297.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l128 128c12.5 12.5 32.8 12.5 45.3 0l256-256z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
        <div>
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-trash action-icon"
            data-icon="trash"
            data-prefix="fas"
            data-testid="delete-button"
            focusable="false"
            onClick={[Function]}
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="App__todo-list__item"
    >
      <div
        className="hstack gap-3"
      >
        <div>
          <div
            className="App__todo-list__item-title"
          >
            Task 6
          </div>
          <div
            className="App__todo-list__item-description"
          >
            <span>
              Description 6
            </span>
            <span>
               (Fri Mar 04 2022)
            </span>
          </div>
        </div>
        <div
          className="ms-auto"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-check-double action-icon"
            data-icon="check-double"
            data-prefix="fas"
            data-testid="complete-button"
            focusable="false"
            onClick={[Function]}
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M374.6 86.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 178.7l-57.4-57.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l80 80c12.5 12.5 32.8 12.5 45.3 0l160-160zm96 128c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 402.7 86.6 297.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l128 128c12.5 12.5 32.8 12.5 45.3 0l256-256z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
        <div>
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-trash action-icon"
            data-icon="trash"
            data-prefix="fas"
            data-testid="delete-button"
            focusable="false"
            onClick={[Function]}
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </div>
  </section>
</main>
`;

exports[`todo list rendered match snapshot loader 1`] = `
<main
  className="App__todo-list"
  onScroll={[Function]}
  style={Object {}}
>
  <div
    className="loader-container d-flex justify-content-center align-items-center fade-in"
    data-testid="loader"
  >
    <div
      className="loader"
    />
  </div>
</main>
`;
